### PR TITLE
[FIX] purchase: prevent traceback when creating PO without company_id

### PR DESCRIPTION
**Pull request by maip-odoo**
When the user remove the company_id and add a line in the purchase order then the issue is generated.

Steps to reproduce:

- Create a new purchase order.
- Remove the company_id in the Purchase order.
- Now add a line in PO. Select a product with the consumable product type.
- Traceback will be generated.

Traceback:
```  File "/home/odoo/odoo/community/odoo/http.py", line 2116, in __call__
    response = request._serve_db()
  File "/home/odoo/odoo/community/odoo/http.py", line 1698, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/home/odoo/odoo/community/odoo/service/model.py", line 134, in retrying
    result = func()
  File "/home/odoo/odoo/community/odoo/http.py", line 1725, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/home/odoo/odoo/community/odoo/http.py", line 1923, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/home/odoo/odoo/community/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/home/odoo/odoo/community/odoo/http.py", line 715, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/home/odoo/odoo/community/addons/web/controllers/dataset.py", line 28, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/odoo/community/addons/web/controllers/dataset.py", line 24, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/odoo/odoo/community/odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/odoo/odoo/community/odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/odoo/community/odoo/models.py", line 6536, in onchange
    todo = [
  File "/home/odoo/odoo/community/odoo/models.py", line 6539, in <listcomp>
    if name not in done and snapshot0.has_changed(name)
  File "/home/odoo/odoo/community/odoo/models.py", line 6322, in has_changed
    return self[name] != record[name]
  File "/home/odoo/odoo/community/odoo/models.py", line 5942, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/home/odoo/odoo/community/odoo/fields.py", line 1155, in __get__
    self.recompute(record)
  File "/home/odoo/odoo/community/odoo/fields.py", line 1365, in recompute
    apply_except_missing(self.compute_value, recs)
  File "/home/odoo/odoo/community/odoo/fields.py", line 1338, in apply_except_missing
    func(records)
  File "/home/odoo/odoo/community/odoo/fields.py", line 1387, in compute_value
    records._compute_field_value(self)
  File "/home/odoo/odoo/community/odoo/models.py", line 4312, in _compute_field_value
    fields.determine(field.compute, self)
  File "/home/odoo/odoo/community/odoo/fields.py", line 99, in determine
    return needle(*args)
  File "/home/odoo/odoo/community/addons/purchase/models/purchase.py", line 1272, in _compute_price_unit_and_date_planned_and_name
    price_unit = seller.currency_id._convert(price_unit, line.currency_id, line.company_id, line.date_order, False)
  File "/home/odoo/odoo/community/odoo/addons/base/models/res_currency.py", line 267, in _convert
    assert company, "convert amount from unknown company"
AssertionError: convert amount from unknown company
```

Applying these changes will resolve this issue.

Sentry-4215579151

### CODE REVIEW

**Comment by maip-odoo** [comment]
@odoo/sentry_reviewers 
Can you please look into this PR?
Thanks!!

**Review by adwid — CHANGES_REQUESTED** [review]
The issue comes from the onchange that triggers the traceback: `:PurchaseOrderLine._onchange_quantity`
Because it uses the fields `.company_id` and `.order_id.company_id` (which are the same) but does not consider that they could be empty, this is the error (in onchange/compute methods, we must consider that required fields could be empty, and we should deal with that)

So, in this use case, we should rather do

```diff
diff --git a/addons/purchase/models/purchase.py b/addons/purchase/models/purchase.py
index 3de58892b986..3c91e2c84870 100644
--- a/addons/purchase/models/purchase.py
+++ b/addons/purchase/models/purchase.py
@@ -1086,6 +1086,7 @@ class PurchaseOrderLine(models.Model):
     def _onchange_quantity(self):
         if not self.product_id or self.invoice_lines:
             return
+        line_company = self.company_id or self.env.company
         params = {'order_id': self.order_id}
         seller = self.product_id._select_seller(
             partner_id=self.partner_id,
@@ -1104,23 +1105,24 @@ class PurchaseOrderLine(models.Model):
                 self.product_id.uom_id._compute_price(self.product_id.standard_price, po_line_uom),
                 self.product_id.supplier_taxes_id,
                 self.taxes_id,
-                self.company_id,
+                line_company,
             )
-            if price_unit and self.order_id.currency_id and self.order_id.company_id.currency_id != self.order_id.currency_id:
-                price_unit = self.order_id.company_id.currency_id._convert(
+            if price_unit and self.order_id.currency_id and line_company.currency_id != self.order_id.currency_id:
+                price_unit = line_company.currency_id._convert(
                     price_unit,
                     self.order_id.currency_id,
-                    self.order_id.company_id,
+                    line_company,
                     self.date_order or fields.Date.today(),
                 )
 
             self.price_unit = price_unit
             return
 
-        price_unit = self.env['account.tax']._fix_tax_included_price_company(seller.price, self.product_id.supplier_taxes_id, self.taxes_id, self.company_id) if seller else 0.0
+        price_unit = self.env['account.tax']._fix_tax_included_price_company(seller.price, self.product_id.supplier_taxes_id, self.taxes_id,
+                                                                             line_company) if seller else 0.0
         if price_unit and seller and self.order_id.currency_id and seller.currency_id != self.order_id.currency_id:
             price_unit = seller.currency_id._convert(
-                price_unit, self.order_id.currency_id, self.order_id.company_id, self.date_order or fields.Date.today())
+                price_unit, self.order_id.currency_id, line_company, self.date_order or fields.Date.today())
 
         if seller and self.product_uom and seller.product_uom != self.product_uom:
             price_unit = seller.product_uom._compute_price(price_unit, self.product_uom)
```

**Comment by maip-odoo** [comment]
@adwid 
Changes are done!!!

File: addons/purchase/models/purchase.py [hunk 1]
```diff
@@ -1104,10 +1105,10 @@ def _onchange_quantity(self):
...
-                self.company_id,
+                line_company,
             )
-            if price_unit and self.order_id.currency_id and self.order_id.company_id.currency_id != self.order_id.currency_id:
-                price_unit = self.order_id.company_id.currency_id._convert(
+            if price_unit and self.order_id.currency_id and line_company.currency_id != self.order_id.currency_id:
+                price_unit = line_company.currency_id._convert(
                     price_unit,
                     self.order_id.currency_id,
                     self.order_id.company_id,
```
Comment (ticodoo) [inline addons/purchase/models/purchase.py:1114]: :raised_eyebrow: you missed one

**Review by ticodoo — COMMENTED** [review]
It seems like a cleaner fix to me would be to add `company_id` to the onchange dependencies ( to both `_onchange_quantity` and `onchange_product_id`) and add `or not company_id` to the `if not self.product_id:` condition in both methods.

i.e.:
```
if not self.product_id or not company_id:
    return
```

this current fix can lead to something like, the company is changed after the product/qty/uom and the price unit can be wrong afterwards since it won't recalculate

**Comment by ticodoo** [comment]
note that this fix works because company_id is required

@robodoo r+

**Pull request closed without merging**